angular-workspace/projects/ni/nimble-angular/pipes/duration.pipe.ts

```diff
@@ -0,0 +1,29 @@
+import { Inject, LOCALE_ID, Pipe, PipeTransform } from '@angular/core';
+import { DurationFormatter } from '@ni/nimble-components/dist/esm/table-column/duration-text/models/duration-formatter';
+
+/**
+ * A pipe that transforms a number of milliseconds (string/number) into a string formatted like "1 day, 2 hr, 3 min, 4.567 sec"
+ */
+@Pipe({
+    name: 'duration'
+})
+export class DurationPipe implements PipeTransform {
+    private readonly durationFormatter;
+
+    public constructor(@Inject(LOCALE_ID) locale: string) {
+        this.durationFormatter = new DurationFormatter(locale);
+    }
+
+    public transform(timeInMilliseconds: string | number | null | undefined): string {
```

Comment:
Note: I made this pipe take both `string` and `number` since that's what the [ElapsedTimePipe does](https://ni.visualstudio.com/DevCentral/_git/Skyline?path=/Web/Workspaces/SystemLinkShared/projects/systemlink-lib-angular/src/pipes/elapsed-time.pipe.ts&version=GBmaster&line=18&lineEnd=19&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents), but I don't feel strongly about it.

I wanted to make it easier for existing `ElapsedTimePipe` usages to switch to this, but in practice it won't be a drop-in replacement, since our duration formatter takes milliseconds as input, whereas ElapsedTimePipe takes seconds.

I also thought about adding a secondary `transformSeconds()` method that took seconds as an input. That would let usages like [in sl-test-results-grid](https://ni.visualstudio.com/DevCentral/_git/Skyline?path=/Web/Workspaces/SystemLinkShared/projects/systemlink-test-results-lib/src/components/sl-test-results-grid/sl-test-results-grid.component.ts&version=GBmaster&line=165&lineEnd=166&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) to just call a single method and not need to multiply by 1000. However once the switchover to `sl-table` is complete and they're using `nimble-table-column-duration-text`/`sl-table-column-duration-text` directly, they'd still need to convert their model values to milliseconds explicitly anyway. So long term I'm not sure a `transformSeconds` method would be that useful.

Reply:
> I made this pipe take both string and number

Looks like Angular's numbery pipes do that too so seems inline with common patterns:
https://angular.io/api/common/DecimalPipe#input-value
https://angular.io/api/common/PercentPipe#input-value
https://angular.io/api/common/CurrencyPipe#input-value

> I also thought about adding a secondary transformSeconds()

That wouldn't actually implement the behavior of the PipeTransform interface right? Think it would have to be a separate pipe that maybe extends / uses this one. But yea even if something like that is useful it would be systemlink-specific and on that side of the fence

Reply:
Correct, the secondary method wouldn't be part of the PipeTransform API. On the SLE side, some code instantiates the pipe directly and calls `transform()` from TypeScript code, so it would work for that purpose.  
But good point that we could add something on the SystemLink side if needed. So I'll keep this as-is.